astropy/utils/misc.py

```diff
@@ -850,3 +850,121 @@ def set_locale(name):
                 yield
             finally:
                 locale.setlocale(locale.LC_ALL, saved)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class ShapedLikeNDArray(object):
+    """Mixin class to provide shape-changing methods.
+
+    The class proper must define abstract properties ``ndim`` and ``shape``,
+    as well as an abstract method ``_replicate``.
+    """
+
+    @abc.abstractproperty
+    def shape(self):
+        """The shape of the instance and underlying arrays."""
+
```

Comment:
perhaps this should have a concerete ``size`` propertry implemented like this:

```
def size(self):
    return int(np.prod(self.shape))
```
so that then size comes for free even if only `shape` is implemented?

For all implementations above you directly use the ``size`` from the data object, so that's just fine.  But this is an added bonus for possible future subclasses.